src/audio/dp_queue.c

```diff
@@ -0,0 +1,351 @@
+		offset -= dp_queue->data_buffer_size;
+	return dp_queue->data_buffer + offset;
+}
+
+static inline
+uint32_t dp_queue_inc_offset(struct dp_queue *dp_queue, uint32_t offset, uint32_t inc)
+{
+	offset += inc;
+	/* wrap around ? keep in mind that offset may be in "double area" */
+	if (offset >= 2*dp_queue->data_buffer_size)
+		offset -= 2*dp_queue->data_buffer_size;
+	return offset;
+}
+
+static inline
+size_t _dp_queue_get_data_available(struct dp_queue *dp_queue)
+{
+	int32_t avail_data =  dp_queue->write_offset - dp_queue->read_offset;
+	/* wrap around ? keep in mind that offset may be in "double area" */
+	if (avail_data < 0)
+		avail_data = 2 * dp_queue->data_buffer_size + avail_data;
+
+	return avail_data;
+
+}
+
+static size_t dp_queue_get_data_available(struct sof_source *source)
+{
+	struct dp_queue *dp_queue = dp_queue_from_source(source);
+
+	return _dp_queue_get_data_available(dp_queue);
+}
+
+static size_t dp_queue_get_free_size(struct sof_sink *sink)
+{
+	struct dp_queue *dp_queue = dp_queue_from_sink(sink);
+
+	return dp_queue->data_buffer_size - _dp_queue_get_data_available(dp_queue);
+}
+
+static int dp_queue_get_buffer(struct sof_sink *sink, size_t req_size,
+			       void **data_ptr, void **buffer_start, size_t *buffer_size)
+{
+	struct dp_queue *dp_queue = dp_queue_from_sink(sink);
+
+	if (req_size > dp_queue_get_free_size(sink))
+		return -ENODATA;
+
+	/* note! a sparse warning will be generated here till
+	 * https://github.com/thesofproject/sof/issues/8006 is implemented
+	 */
+	*data_ptr = dp_queue_get_pointer(dp_queue, dp_queue->write_offset);
+	*buffer_start = dp_queue->data_buffer;
+	*buffer_size = dp_queue->data_buffer_size;
+
+	/* no need to invalidate cache - buffer is to be written only */
+	return 0;
+}
+
+static int dp_queue_commit_buffer(struct sof_sink *sink, size_t commit_size)
+{
+	struct dp_queue *dp_queue = dp_queue_from_sink(sink);
+
+	if (commit_size) {
+		dp_queue_writeback_shared(dp_queue,
+					  dp_queue_get_pointer(dp_queue, dp_queue->write_offset),
+					  commit_size);
+
+		/* move write pointer */
+		dp_queue->write_offset =
+				dp_queue_inc_offset(dp_queue, dp_queue->write_offset, commit_size);
+	}
+
+	return 0;
+}
+
+static int dp_queue_get_data(struct sof_source *source, size_t req_size,
+			     void **data_ptr, void **buffer_start, size_t *buffer_size)
+{
+	struct dp_queue *dp_queue = dp_queue_from_source(source);
+
+	if (req_size > dp_queue_get_data_available(source))
+		return -ENODATA;
+
+	/*
+	 * note! a sparse warning will be generated here till
+	 * https://github.com/thesofproject/sof/issues/8006 is implemented
+	 */
+	*buffer_start = dp_queue->data_buffer;
+	*buffer_size = dp_queue->data_buffer_size;
+	*data_ptr = dp_queue_get_pointer(dp_queue, dp_queue->read_offset);
+
+	/* clean cache in provided data range */
+	dp_queue_invalidate_shared(dp_queue, *data_ptr, req_size);
+
+	return 0;
+}
+
+static int dp_queue_release_data(struct sof_source *source, size_t free_size)
```

Comment:
a static function, so not too important, but maybe a name like "update_read" would be clearer - we don't really release data, we release buffer space